File: app/src/main/java/com/robotnec/budget/core/service/aggregation/impl/AggregationServiceImpl.java
package com.robotnec.budget.core.service.aggregation.impl;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.annimon.stream.function.Function;
import com.robotnec.budget.core.domain.Currency;
import com.robotnec.budget.core.domain.MoneyAmount;
import com.robotnec.budget.core.domain.operation.Transaction;
import com.robotnec.budget.core.service.aggregation.AggregationService;

import org.threeten.bp.LocalDateTime;

import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.SortedMap;
import java.util.TreeMap;

/**
 * @author zak <zak@swingpulse.com>
 */
public class AggregationServiceImpl implements AggregationService {

    @Override
    public TransactionAggregation aggregate(List<Transaction> transactions, Resolution resolution) {
        SortedMap<TimeSpan, List<Transaction>> aggregatedMap = new TreeMap<>();
        Map<TimeSpan, MoneyAmount> sums = new HashMap<>();
        if (transactions.isEmpty()) {
            return TransactionAggregationImpl.from(aggregatedMap, sums);
        }

        Comparator<Transaction> comparator = (o1, o2) -> o1.getDate().compareTo(o2.getDate());
        Function<Transaction, LocalDateTime> date = Transaction::getDate;
        LocalDateTime minDate = Stream.of(transactions)
                .min(comparator)
                .map(date)
                .get();
        LocalDateTime maxDate = Stream.of(transactions)
                .max(comparator)
                .map(date)
                .get();

        TimeSpans timeSpans = TimeSpans.dyDays(minDate, maxDate);

        Collections.sort(transactions, comparator);

        for (TimeSpan timeSpan : timeSpans) {
            List<Transaction> values = Stream.of(transactions)
                    .filter(t -> timeSpan.isInSpan(t.getDate()))
                    .collect(Collectors.toList());
            MoneyAmount sum = Stream.of(transactions)
                    .map(Transaction::getAmount)
                    .reduce(MoneyAmount.Companion.of(0, Currency.UAH), MoneyAmount::add);
            if (!values.isEmpty()) {
                aggregatedMap.put(timeSpan, values);
                sums.put(timeSpan, sum);
            }
        }

        return TransactionAggregationImpl.from(aggregatedMap, sums);
    }
}

File: app/src/main/java/com/robotnec/budget/core/service/aggregation/impl/AggregationServiceImpl.kt
package com.robotnec.budget.core.service.aggregation.impl

import com.robotnec.budget.core.domain.operation.Transaction
import com.robotnec.budget.core.service.aggregation.AggregationService
import org.joda.money.CurrencyUnit
import org.joda.money.Money
import java.util.*

/**
 * @author zak zak@swingpulse.com>
 */
class AggregationServiceImpl : AggregationService {

    override fun aggregate(transactions: List<Transaction>, resolution: AggregationService.Resolution): TransactionAggregation {
        val aggregatedMap = TreeMap<TimeSpan, List<Transaction>>()
        val sums = HashMap<TimeSpan, Money>()

        if (transactions.isEmpty()) {
            return TransactionAggregationImpl.from(aggregatedMap, sums)
        }

        val minDate = transactions.map { it.date }.min()
        val maxDate = transactions.map { it.date }.max()

        if (minDate != null && maxDate != null) {
            val timeSpans = TimeSpans.byDays(minDate, maxDate)

            transactions.sortedBy { it.date }

            for (timeSpan in timeSpans) {
                val values = transactions
                        .filter { (_, _, date) -> timeSpan.isInSpan(date) }
                        .toList()

                val sum = transactions
                        .map { it.amount }
                        .fold(Money.of(CurrencyUnit.USD, 0.0)) { obj, other -> obj + other }

                if (!values.isEmpty()) {
                    aggregatedMap[timeSpan] = values
                    sums[timeSpan] = sum
                }
            }
        }

        return TransactionAggregationImpl.from(aggregatedMap, sums)
    }
}
